File: 1026-maximum-difference-between-node-and-ancestor/1026-maximum-difference-between-node-and-ancestor.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxAncestorDiff(self, root: Optional[TreeNode]) -> int:
        res = 0
        
        def dfs(root: Optional[TreeNode], cur_min: int, cur_max: int) -> None:
            nonlocal res
            if not root:
                return
            cur_min = min(cur_min, root.val)
            cur_max = max(cur_max, root.val)
            res = max(res, abs(root.val-cur_min), abs(root.val-cur_max))
            dfs(root.left, cur_min, cur_max)
            dfs(root.right, cur_min, cur_max)
            
        dfs(root, root.val, root.val)
        return res

File: 1026-maximum-difference-between-node-and-ancestor/1026-maximum-difference-between-node-and-ancestor.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function maxAncestorDiff(root: TreeNode | null): number {
    return dfs(root, root.val, root.val, [0])
    
    function dfs(root: TreeNode, min: number, max: number, res: number[]) {
        if (!root) return
        min = Math.min(min, root.val)
        max = Math.max(max, root.val)
        res[0] = Math.max(res[0], Math.abs(root.val-min), Math.abs(root.val-max))
        dfs(root.left, min, max, res)
        dfs(root.right, min, max, res)
        return res[0]
    }
};